path.go

```diff
@@ -0,0 +1,133 @@
+)
+
+const (
+
+	// ClusterHeader set to "<lcluster>" on a request is an alternative to accessing the
+	// cluster via /clusters/<lcluster>. With that the <lcluster> can be access via normal kube-like
+	// /api and /apis endpoints.
+	ClusterHeader = "X-Kubernetes-Cluster"
+
+	separator = ":"
+)
+
+// Path represents a logical cluster. A logical cluster is
+//  1. a (part of) etcd prefix to store objects in that cluster
+//  2. a (part of) a http path which serves a Kubernetes-cluster-like API with
+//     discovery, OpenAPI and the actual API groups.
+//  3. a value in metadata.clusterName in objects from cross-workspace list/watches,
+//     which is used to identify the logical cluster.
+//
+// A logical cluster is a colon separated list of words. In other words, it is
+// like a path, but with colons instead of slashes.
+type Path struct {
+	value string
+}
+
+// New returns a Path from a string.
+func New(value string) Path {
+	return Path{value}
+}
+
+// NewValidated returns a Path from a string and whether it is a valid logical cluster.
+// A valid logical cluster returns true on IsValid().
+func NewValidated(value string) (Path, bool) {
+	n := Path{value}
+	return n, n.IsValid()
+}
+
+// Empty returns true if the logical cluster value is unset.
+func (n Path) Empty() bool {
+	return n.value == ""
+}
+
+// Path returns a path segment for the logical cluster to access its API.
+func (n Path) Path() string {
+	return path.Join("/clusters", n.value)
+}
+
+// String returns the string representation of the logical cluster name.
+func (n Path) String() string {
+	return n.value
+}
+
+// Parent returns the parent logical cluster name of the given logical cluster name.
+func (n Path) Parent() (Path, bool) {
+	parent, _ := n.Split()
+	return parent, parent.value != ""
+}
+
+// Split splits logical cluster immediately following the final colon,
+// separating it into a parent logical cluster and name component.
+// If there is no colon in path, Split returns an empty logical cluster name
+// and name set to path.
+func (n Path) Split() (parent Path, name string) {
+	i := strings.LastIndex(n.value, separator)
+	if i < 0 {
+		return Path{}, n.value
+	}
+	return Path{n.value[:i]}, n.value[i+1:]
+}
+
+// Base returns the last component of the logical cluster name.
+func (n Path) Base() string {
+	_, name := n.Split()
+	return name
+}
+
+// Join joins a parent logical cluster name and a name component.
+func (n Path) Join(name string) Path {
+	if n.value == "" {
+		return Path{name}
+	}
+	return Path{n.value + separator + name}
+}
+
+func (n Path) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&n.value)
+}
+
+func (n *Path) UnmarshalJSON(data []byte) error {
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	n.value = s
+	return nil
+}
+
+func (n Path) HasPrefix(other Path) bool {
+	return strings.HasPrefix(n.value, other.value)
+}
+
+const lclusterNameFmt string = "[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?"
```

Comment:
shouldn't we do validation on the `Name` type as well?

Reply:
yes, in the same way with a `IsValid` func.

Reply:
done